src/Bicep.Core/Semantics/ParamsSemanticModel.cs

```diff
@@ -39,10 +43,79 @@ public ParamsSemanticModel(BicepParamFile bicepParamFile, ImmutableArray<IDiagno
+            this.DeclarationsByAssignment = InitializeAssignmentToDeclarationDictionary();
             // lazy load single use diagnostic set
             this.AllDiagnostics = new Lazy<ImmutableArray<IDiagnostic>>(() => GetAllDiagnostics());
         }
 
+        public ImmutableArray<IDiagnostic> GetAllDiagnostics()
+            => BicepParamFile.ProgramSyntax.GetParseDiagnostics()
+                .Concat(ParamsTypeManager.GetAllDiagnostics())
+                .Concat(this.CompilationLoadDiagnostics)
+                .Concat(GetAdditionalSemanticDiagnostics()).ToImmutableArray();
+            
+        /// <summary>
+        /// Gets the file that was compiled.
+        /// </summary>
+        public ParamFileSymbol Root => this.ParamBinder.ParamFileSymbol;
+
+        public ParameterSymbol? GetParameterDeclaration(ParameterAssignmentSymbol parameterAssignmentSymbol)
+        {
+            return this.DeclarationsByAssignment[parameterAssignmentSymbol];
+        }
+
+        public ParameterAssignmentSymbol? GetParameterAssignment(ParameterSymbol parameterSymbol)
+        {
+            if (this.AssignmentsByDeclaration is null)
+            {
+                return null;
+            }
+            return this.AssignmentsByDeclaration[parameterSymbol];
+        }
+
+        private ImmutableDictionary<ParameterSymbol, ParameterAssignmentSymbol?>? InitializeDeclarationToAssignmentDictionary()
+        {
+            if (this.BicepCompilation is null)
+            {
+                return null;
+            }
+
+            var assignmentsByDeclaration = this.BicepCompilation.GetEntrypointSemanticModel().Root.ParameterDeclarations.ToDictionary(x => x, _ => (ParameterAssignmentSymbol?)null);
+            var parameterAssignments = BicepParamFile.ProgramSyntax.Children.OfType<ParameterAssignmentSyntax>().Where(x => this.ParamBinder.GetSymbolInfo(x) is not null);
+            var assignmentsBySymbolName = parameterAssignments.ToDictionary(x => x.Name.IdentifierName, LanguageConstants.IdentifierComparer);
+
+            foreach (var declaration in assignmentsByDeclaration.Keys)
+            {
+                if (assignmentsBySymbolName.TryGetValue(declaration.Name, out var parameterAssignmentSyntax))
+                {
+                    assignmentsByDeclaration[declaration] = (ParameterAssignmentSymbol?)this.ParamBinder.GetSymbolInfo(parameterAssignmentSyntax);
+                }
+            }
+            return assignmentsByDeclaration.ToImmutableDictionary();
+        }
+
+        private ImmutableDictionary<ParameterAssignmentSymbol, ParameterSymbol?> InitializeAssignmentToDeclarationDictionary()
+        {
+            var declarationsByAssignment = this.ParamBinder.ParamFileSymbol.ParameterAssignmentSymbols.ToDictionary(x => x, _ => (ParameterSymbol?)null);
+            if (this.BicepCompilation is null)
+            {
+                return declarationsByAssignment.ToImmutableDictionary();
```

Comment:
I'd suggest doing the null check at the beginning and return `ImmutableDictionary<ParameterAssignmentSymbol, ParameterSymbol?>.Empty`